MericCelik/assignment3/Main.java

```diff
@@ -0,0 +1,106 @@
+import java.util.ArrayList;
+import java.util.HashSet;
+
+/**Solution class that uses Dictionary class.
+ * @author AycaMericCelik*/
+
+public class Main {
+    public static void main(String[] args){
+        String[][] grids = {"aar".split(""), "tcd".split("")};
+        String[] words = {"CAR", "CARD", "CART", "CAT"};
+        Dictionary dict = new Dictionary(words);
+        TrieBasedDictionary dict2  = new TrieBasedDictionary(words);
+
+        System.out.println(findWords(grids, dict));
+        System.out.println(TrieBasedDictionary.findWords(grids,dict2));
+
+    }
+
+    /** Finds all words that can be created using grids. Calls buildAllWords method for each starting grid.
+     * It creates zero-filled visited matrix for each call of buildAllWords method. It passes current grids
+     * coordinates, grids matrix and the dictionary.
+     * @param dict : The dictionary that contains some words and prefixes.
+     * @param grids :The grids matrix to construct words from its elements.
+     * @return set of possible words(no duplicates)
+     * @throws NullPointerException when the parameters are invalid.
+     * */
+    public static HashSet<String> findWords(String[][] grids, Dictionary dict){
+
+        if(grids == null || dict ==null)
+            throw new NullPointerException("Invalid parameter!");
```

Comment:
Very nice! How about including which one of them is null in the error message?

Reply:
I tried to refactor the code to avoid duplicates. Now it takes two other parameter named "triedict" and "isTrieBased". It throws exception in these cases:
- isTrieBased is false, dict is null,
- isTrieBased is true, triedict is null.

In other situations, the algorithm does not use the null dictionary because of the isTrieBased, so it is not a problem. I hope they are clean and healthy ways to do, both refactoring and handling exceptions. :blush: